lib/livebook/web_socket.ex

```diff
@@ -0,0 +1,96 @@
+defmodule Livebook.WebSocket do
+  @moduledoc false
+
+  alias Livebook.WebSocket.Client
+  alias LivebookProto.{Request, SessionRequest}
+
+  defmodule Connection do
+    defstruct [:conn, :websocket, :ref]
+
+    @type t :: %__MODULE__{
+            conn: Client.conn(),
+            websocket: Client.websocket(),
+            ref: Client.ref()
+          }
+  end
+
+  @type proto :: SessionRequest.t()
+
+  @typep header :: {String.t(), String.t()}
+  @typep headers :: list(header())
+
+  @doc """
+  Connects with the WebSocket server for given URL and headers.
+  """
+  @spec connect(String.t(), headers()) ::
+          {:ok, Connection.t(), :connected | {atom(), proto()}}
+          | {:error, Connection.t(), String.t() | LivebookProto.Error.t()}
+  def connect(url, headers \\ []) do
+    with {:ok, conn, ref} <- Client.connect(url, headers) do
+      conn
+      |> Client.receive(ref)
+      |> handle_receive(ref)
+    end
+  end
+
+  @doc """
+  Disconnects the given WebSocket client.
+  """
+  @spec disconnect(Connection.t()) :: :ok
+  def disconnect(%Connection{} = connection) do
+    Client.disconnect(connection.conn, connection.websocket, connection.ref)
+  end
+
+  @doc """
+  Sends a request to the given server.
+  """
+  @spec send_request(Connection.t(), proto()) ::
+          {:ok, Connection.t()}
+          | {:error, Connection.t(), Client.ws_error() | Client.mint_error()}
+  def send_request(%Connection{} = connection, %{__struct__: SessionRequest} = session_request) do
```

Comment:
```suggestion
  def send_request(%Connection{} = connection, %SessionRequest{} = session_request) do
```

But if we can't retrieve the atom `:session` programatically, I think I would do this instead:

    send_request(connection, {:session, SessionRequest.new!(...)})